File: kotlin/triangle/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {

    val sides = listOf(a, b, c)
    val uniq = sides.distinct()
    val sorted = sides.map { it.toDouble() }.sorted()

    init {
        require(sorted.all { it > 0}.and(sorted[0] + sorted[1] >= sorted[2]))
    }

    val isEquilateral: Boolean =  uniq.size == 1
    val isIsosceles: Boolean = uniq.size <= 2
    val isScalene: Boolean = uniq.size == 3
}

File: typescript/triangle/triangle.ts
export class Triangle {
  private sorted: number[]
  private set: Set<number>;

  constructor(...sides: number[]) {
    this.sorted = sides.sort((a,b) => a - b)
    this.set = new Set(sides);
  }

  valid = () => this.sorted[0] > 0 && this.sorted[0] + this.sorted[1] >= this.sorted[2]

  get isEquilateral() {
    return this.valid() && this.set.size === 1
  }

  get isIsosceles() {
    return this.valid() && this.set.size <= 2
  }

  get isScalene() {
    return this.valid() && this.set.size === 3
  }
}
